### PR TITLE
Memcache proxy: read all the input buffer, increase buffer size, add tests for max me…

### PR DESCRIPTION
**Pull request by rdcastro**
This change:
- Makes sure that we read all the input by using io.ReadFull() - this fixes the issue of not reading beyond the default buffer size (4096)
- Increases buffer size to help reading larger memcache values
- Add unit tests, incl. max memcache key size

### CODE REVIEW

**Comment by oierw** [comment]
I think this demonstrates one of a class of bugs (insufficient reading of input) in the memcache proxy.
Class 1: We may not read enough data to work possibly. The most obvious place this would happen is in serveConn's loop. We can imagine that the call to s.in.Read('\n') doesn't get a full set command and is truncated at "se". All logic from that point on won't work.
Class 2: In this PR we try to read more data, but sadly the protocol has another entry *after* the data to set. We never read it (and will reply if told noreply).

The memcache text protocol is \r\n delimited. I think we have to change how we parse to fix it.

I think we need a helper function that will read to s.in until it reaches a '\n', no matter the length. The buffer it puts data in only has to be 1,000,286[1] bytes. If the message is longer, we should read it until we get a \n and return an error.
An alternative would be to rewrite chunks of code to do whitespace separation to parse the protocol, but that seems harder.

1: The longest acceptable input is "replace $key $flags $exptime $bytes noreply\r\n"
The non-variable input in that string is 21 characters.
Key is allowed to be 250 characters.
Flags is a 16 bit number in decimal. 5 characters
Exptime is unbounded in the spec, but we decode it to a 32 bit number, in decimal. 10 characters
Bytes is 1,000,000 max.
That gives us 1,000,286 bytes max

**Comment by rdcastro** [comment]
Note that serveConn loop uses ReadBytes (not Read). Unlike Read, ReadBytes guarantees that the buffer is read until the delimiter (unless there is an error, which is handled by the code).

File: memcache_proxy/dtog.go [hunk 1]
```diff
@@ -371,6 +379,16 @@ func unpackFirstStorageCmdArgs(args ...[]byte) (key []byte, flags uint32,
...
 	bytes = uint32(bytesUpc)
+	// Make sure we only accept up to 1000000 bytes to read (max value size
+	// for App Engine cache). Other parts of the code will allocate memory
+	// based on this number and this will only fail once we actually hit
+	// App Engine Memcache.
+	if bytes > maxMemcacheValueSize {
+		err = clientError{"bad command line format",
+			fmt.Sprintf("Value size should not be greater than %d, but got %d.",
+				maxMemcacheValueSize, bytes)}
+		return
```
Comment (JustinBeckwith) [inline memcache_proxy/dtog.go]: Why the naked return?  Are we passing errors up?

File: memcache_proxy/dtog.go [hunk 2]
```diff
@@ -371,6 +379,16 @@ func unpackFirstStorageCmdArgs(args ...[]byte) (key []byte, flags uint32,
 		return
 	}
 	bytes = uint32(bytesUpc)
+	// Make sure we only accept up to 1000000 bytes to read (max value size
+	// for App Engine cache). Other parts of the code will allocate memory
+	// based on this number and this will only fail once we actually hit
+	// App Engine Memcache.
+	if bytes > maxMemcacheValueSize {
+		err = clientError{"bad command line format",
```
Comment (waprin) [inline memcache_proxy/dtog.go]: Is "bad command line format" actually the correct error message, or was it an artifact of the memcache server choking while reading a packet truncated by the server? looks to me like the error should be more directly stating the problem MEMCACHED_E2BIG:
http://docs.libmemcached.org/client_errors/MEMCACHED_E2BIG.html#MEMCACHED_E2BIG . Also, whatever part of the error message does state the actual problem, we need to be sure that part gets surfaced to cloud logging.

**Comment by eobrain** [comment]
Is it possible to add a test that fails without this fix?

File: memcache_proxy/dtog.go (hunk 1 above)
Comment (rdcastro) [inline memcache_proxy/dtog.go]: err is an output param and is being assigned here, so we're good. The other output params don't need to be set.

**Comment by rdcastro** [comment]
The 1Mb test I added fails without this fix. I had added initially a test with a 4097 byte value, which reproed the problem, but the 1Mb is enough for it.

File: memcache_proxy/dtog.go (hunk 2 above)
Comment (rdcastro) [inline memcache_proxy/dtog.go]: Done, thanks.

File: memcache_proxy/dtog.go [hunk 3]
```diff
@@ -33,6 +33,12 @@ import (
 	pb "google.golang.org/appengine/notreallyinternal/memcache"
 )
 
+const (
+	// Max App Engine Memcache value size (10^6 bytes).
+	// https://cloud.google.com/appengine/docs/python/memcache/
+	maxMemcacheValueSize = 1000000
```
Comment (eobrain) [inline memcache_proxy/dtog.go]: Note that although the Python API enforces 10^6 max (1 MB), the underlying system can handle 2^20 (1 MiB) of key+value+overhead.  This is exposed, for example, in the [Go API docs][1]: "maximum size of a cached data value is 1 MiB (2^20 bytes) minus the size of the key minus an implementation-dependent overhead, which is approximately 73 bytes"

[1]: https://cloud.google.com/appengine/docs/go/memcache/#Go_Limits

Comment (rdcastro) [inline memcache_proxy/dtog.go]: Interesting. I'm tempted to focus on the 4K bug here and tackling this validation separately, as it might require doing proper testing across all languages to make sure we got it right. Let me know your thoughts.

Comment (rdcastro) [inline memcache_proxy/dtog.go]: Java has the same limits as Go, so I think it's better to do this properly in a separate CL.

**Comment by eobrain** [comment]
LGTM

**Comment by waprin** [comment]
LGTM, big thanks!

**Comment by oierw** [comment]
LGTM
My protocol level issues were because I was somehow was looking at a memcache text protocol spec where the value was inline.
It seems we are doing a different spec where there is a \r\n between the set request and the contents to set.
This looks like it plugs that hole exactly correctly.
